include/sys/arc_impl.h

```diff
@@ -176,6 +177,212 @@ typedef struct l1arc_buf_hdr {
 	abd_t			*b_pabd;
 } l1arc_buf_hdr_t;
 
+enum {
+	L2ARC_DEV_HDR_EVICT_FIRST = (1 << 0)	/* mirror of l2ad_first */
+};
+
+/*
+ * Pointer used in persistent L2ARC (for pointing to log blocks & ARC buffers).
+ */
+typedef struct l2arc_log_blkptr {
+	uint64_t	lbp_daddr;	/* device address of log */
+	/*
+	 * lbp_prop is the same format as the blk_prop in blkptr_t:
+	 *	* logical size (in sectors)
+	 *	* physical (compressed) size (in sectors)
+	 *	* compression algorithm (we always LZ4-compress l2arc logs)
+	 *	* checksum algorithm (used for lbp_cksum)
+	 *	* object type & level (unused for now)
+	 */
+	uint64_t	lbp_prop;
+	zio_cksum_t	lbp_cksum;	/* fletcher4 of log */
+} l2arc_log_blkptr_t;
+
+/*
+ * The persistent L2ARC device header.
+ * Byte order of magic determines whether 64-bit bswap of fields is necessary.
+ */
+typedef struct l2arc_dev_hdr_phys {
+	uint64_t	dh_magic;	/* L2ARC_DEV_HDR_MAGIC */
+	zio_cksum_t	dh_self_cksum;	/* fletcher4 of fields below */
+
+	/*
+	 * Global L2ARC device state and metadata.
+	 */
+	uint64_t	dh_spa_guid;
+	uint64_t	dh_alloc_space;		/* vdev space alloc status */
+	uint64_t	dh_flags;		/* l2arc_dev_hdr_flags_t */
+
+	/*
+	 * Start of log block chain. [0] -> newest log, [1] -> one older (used
+	 * for initiating prefetch).
+	 */
+	l2arc_log_blkptr_t	dh_start_lbps[2];
+
+	const uint64_t	dh_pad[44];		/* pad to 512 bytes */
+} l2arc_dev_hdr_phys_t;
+
+/*
+ * A single ARC buffer header entry in a l2arc_log_blk_phys_t.
+ */
+typedef struct l2arc_log_ent_phys {
+	dva_t			le_dva;	/* dva of buffer */
+	uint64_t		le_birth;	/* birth txg of buffer */
+	zio_cksum_t		le_freeze_cksum;
+	/*
+	 * le_prop is the same format as the blk_prop in blkptr_t:
+	 *	* logical size (in sectors)
+	 *	* physical (compressed) size (in sectors)
+	 *	* compression algorithm
+	 *	* checksum algorithm (used for b_freeze_cksum)
+	 *	* object type & level (used to restore arc_buf_contents_t)
+	 */
+	uint64_t		le_prop;
+	uint64_t		le_daddr;	/* buf location on l2dev */
+	boolean_t		le_protected;
+	const uint32_t		le_pad[13];	/* resv'd for future use */
+} l2arc_log_ent_phys_t;
+
+/*
+ * These design limits give us the following metadata overhead (before
+ * compression):
+ *	avg_blk_sz	overhead
+ *	1k		12.51 %
+ *	2k		 6.26 %
+ *	4k		 3.13 %
+ *	8k		 1.56 %
+ *	16k		 0.78 %
+ *	32k		 0.39 %
+ *	64k		 0.20 %
+ *	128k		 0.10 %
+ * Compression should be able to squeeze these down by about a factor of 2x.
+ */
+#define	L2ARC_LOG_BLK_SIZE			(128 * 1024)	/* 128k */
+#define	L2ARC_LOG_BLK_HEADER_LEN		(128)
+#define	L2ARC_LOG_BLK_ENTRIES			/* 1023 entries */	\
+	((L2ARC_LOG_BLK_SIZE - L2ARC_LOG_BLK_HEADER_LEN) /		\
+	sizeof (l2arc_log_ent_phys_t))
+/*
+ * Maximum amount of data in an l2arc log block (used to terminate rebuilding
+ * before we hit the write head and restore potentially corrupted blocks).
+ * Since the max recordsize/volblocksize is 128Kbytes, the max payload would
+ * be 128Kbytes * 1023 entries.
```

Comment:
I did notice that. @skiselkov argues in the comments that "we need the L2ARC device to at least be able to hold 3 full log blocks" (Line 275). In this case, L2ARC_PERSIST_MIN_SIZE would increase to 3*(16MB*1023entries)=49GB. So if the L2ARC is less than 49GB it will not be made persistent.

Reply:
You could make it use 128k just for the tests, as the codecov is mostly the same regardless of 128k vs 16m...

And thanks @behlendorf for taking the time to review this :)

Reply:
@Ornias1993 how can I make it use 128k just for the tests?

Reply:
Well I don't know for sure... and it would sorta invalidate the test results...

Maybe just use a sparse test volume?

Reply:
I am not convinced that the minimum size of L2ARC should be 3x the amount of the max L2ARC log block payload. Tests run fine even if it is equals the max L2ARC log block payload.

Reply:
I am not convinced that L2ARC_PERSIST_MIN_SIZE should be 3x the size of the maximum payload. Testing reveals the code works fine even if it equals the size of the maximum payload. The only implication I see is that due to the way log blocks are restored (via 2 alternating linked lists, for performance reasons), it expects at least 2 log blocks (with their payload) to be present in L2ARC. If L2ARC_PERSIST_MIN_SIZE equals the max payload of just 1 log block, then the latest one is restored properly while the second one is not and as its checksum is not valid (line 9382 in module/zfs/arc.c).

Reply:
If we move the "end of list detection" in l2arc_rebuild() before the call to l2arc_log_blk_read() we can actually get away with that.  In this case even when L2ARC_PERSIST_MIN_SIZE equals the max payload of just 1 log block, only the most recent log block is restored and the rebuild stops after that.

Reply:
Pushed a new commit ("fix max log block payload") addressing this issue.

Reply:
I reverted that commit since it leads to leaking of abt_t objects.

Reply:
I figured out the leak, pushed new commit.